discovery-provider/src/tasks/cache_user_balance.py

```diff
@@ -0,0 +1,148 @@
+#
+# - In populate_track_metadata, look up User_Balance entry in db.
+#       If it doesn't exist, return 0, persist a User_Balance row with 0,
+#       & enqueue in Redis the user ID for later balance lookup.
+# - On track get endpoints, if current_user exists, enqueue balance lookup
+#       in Redis.
+#
+# In this recurring task:
+#   - Get all enqueued user balance refresh requests. These are stored in a set, so
+#       we don't worry about deduping.
+#   - If a given balance is either
+#        a) new (created_at == updated_at)
+#        b) not new, but stale: last updated prior to (now - threshold)
+#     we look up said users, adding User_Balance rows, and removing them from Redis.
+#
+#     enqueued User Ids in Redis that are *not* ready to be refreshed yet are left in the queue
+#     for later.
+
+def refresh_user_ids(redis, db, token_contract, eth_web3):
+    # List users in Redis set, balances decoded as strings
+    redis_user_ids = redis.smembers(REDIS_PREFIX)
+    redis_user_ids = [int(user_id.decode()) for user_id in redis_user_ids]
+
+    if not redis_user_ids:
+        return
+
+    logger.info(f"cache_user_balance.py | Starting refresh with {len(redis_user_ids)} user_ids: {redis_user_ids}")
+
+    with db.scoped_session() as session:
+        query = ((
+            session.query(UserBalance)
+        ).filter(
+            UserBalance.user_id.in_(redis_user_ids)
+        ).all())
+
+        # Balances from current user lookup may
+        # not be present in the db, so make those
+        not_present_set = {user_id for user_id in redis_user_ids} - {user.user_id for user in query}
+        new_balances = [UserBalance(user_id=user_id, balance=0) for user_id in not_present_set]
+        if new_balances:
+            session.add_all(new_balances)
+            logger.info(f"cache_user_balance.py | adding new users: {not_present_set}")
+
+        # Filter only user_balances that still need refresh
+        # This includes new balances
+        needs_refresh = list(filter(does_user_balance_need_refresh, query))
+        needs_refresh += new_balances
+
+        logger.info(f"cache_user_balance.py | needs refresh: {needs_refresh}")
+
+        # map user_id -> user_balance
+        needs_refresh_map = {user.user_id: user for user in needs_refresh}
+
+        # Grab the users we need to refresh
+        user_query = ((
+            session.query(User)
+        ).filter(
+            User.user_id.in_(needs_refresh_map.keys()),
+            User.is_current == True,
+        )).all()
+
+        logger.info(f"cache_user_balance.py | fetching for {len(user_query)} users: {needs_refresh_map.keys()}")
+
+        # Fetch balances
+        for user in user_query:
+            wallet, user_id = user.wallet, user.user_id
+            wallet = eth_web3.toChecksumAddress(wallet)
+
+            # get balance
+            balance = token_contract.functions.balanceOf(wallet).call()
```

Comment:
this might take a super long time since it has to do all of these sequentially right? is there a way to maybe make these requests in parallel? i know python makes this a bummer. if we have like 300 users and each takes 40ms, that's already 12s

Reply:
@hareeshnagaraj did you run into this at all? asyncio or aiohttp or some similar library could work. or we could leverage celery more and have multiple workers

Reply:
there is a way to parallelize this if that's what you're getting at? w/concurrent.futures

Reply:
there are refs in index_network_peers of this for example

Reply:
I can look into this - this will definitely take some time especially when we're first getting bootstrapped (it's ~400ms per call to getBalance). I'm not sure how much of an issue it is if a single run through the task is long running though

Reply:
this is a good point, and worth benchmarking on prod like we've been discussing